fastcrypto-proving/benches/proving.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use ark_bls12_381::{Bls12_381, Fr as BlsFr};
+use ark_crypto_primitives::SNARK;
+use ark_ec::PairingEngine;
+use ark_ff::PrimeField;
+use ark_groth16::Groth16;
+use ark_relations::{
+    lc,
+    r1cs::{ConstraintSynthesizer, ConstraintSystemRef, SynthesisError},
+};
+use criterion::{
+    criterion_group, criterion_main, measurement::Measurement, BenchmarkGroup, BenchmarkId,
+    Criterion, SamplingMode,
+};
+
+#[derive(Copy, Clone)]
+struct DummyCircuit<F: PrimeField> {
+    pub a: Option<F>,
+    pub b: Option<F>,
+    pub num_variables: usize,
+    pub num_constraints: usize,
+}
+
+impl<F: PrimeField> ConstraintSynthesizer<F> for DummyCircuit<F> {
+    fn generate_constraints(self, cs: ConstraintSystemRef<F>) -> Result<(), SynthesisError> {
+        let a = cs.new_witness_variable(|| self.a.ok_or(SynthesisError::AssignmentMissing))?;
+        let b = cs.new_witness_variable(|| self.b.ok_or(SynthesisError::AssignmentMissing))?;
+        let c = cs.new_input_variable(|| {
+            let a = self.a.ok_or(SynthesisError::AssignmentMissing)?;
+            let b = self.b.ok_or(SynthesisError::AssignmentMissing)?;
+
+            Ok(a * b)
+        })?;
+
+        for _ in 0..(self.num_variables - 3) {
+            let _ = cs.new_witness_variable(|| self.a.ok_or(SynthesisError::AssignmentMissing))?;
+        }
+
+        for _ in 0..self.num_constraints - 1 {
+            cs.enforce_constraint(lc!() + a, lc!() + b, lc!() + c)?;
+        }
+
+        cs.enforce_constraint(lc!(), lc!(), lc!())?;
+
+        Ok(())
+    }
+}
+
+fn bench_prove<F: PrimeField, E: PairingEngine<Fr = F>, M: Measurement>(
+    grp: &mut BenchmarkGroup<M>,
+) {
+    static VARIABLES: [usize; 5] = [16, 32, 64, 128, 256];
+
+    for size in VARIABLES.iter() {
+        let rng = &mut ark_std::test_rng();
+        let c = DummyCircuit::<F> {
+            a: Some(<F>::rand(rng)),
+            b: Some(<F>::rand(rng)),
+            num_variables: *size,
+            num_constraints: 65536,
```

Comment:
how did you come up to this number?

Reply:
I've actually refactored the test to focus on the # of constraints, which the proof timing scales with. The number of variables makes little sense considering we only provide one input.